bittensor/_axon/__init__.py

```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
             
         if modality == bittensor.proto.Modality.TENSOR:
             sample_input = torch.rand(1,1,1)
-            forward_callback(pubkey,sample_input)
\ No newline at end of file
+            forward_callback(pubkey,sample_input)
+
+class AuthInterceptor(grpc.ServerInterceptor):
+    def __init__(self,keypair, key:str = 'Bittensor'):
+        r""" Creates a new server interceptor that authenticates incoming messages from passed arguments.
+        Args:
+            key (str, `optional`):
+                 key for authentication header in the metadata (default= Bittensor)
+            keypair (:obj:`callable`, `optional`): 
+                Substrate interface keypair object
+        """
+        self._valid_metadata = ('rpc-auth-header', key)
+        self.keypair = keypair
+        self.nounce_dic = {}
+
+        def deny(_, context):
+            context.abort(grpc.StatusCode.UNAUTHENTICATED, 'Invalid key')
+
+        self._deny = grpc.unary_unary_rpc_method_handler(deny)
+
+    def intercept_service(self, continuation, handler_call_details):
+        r""" Authentication between bittensor nodes. A signature is created using
+            publickey and encodes the time at which the signature was created.
+        """
+        meta = handler_call_details.invocation_metadata
+        if meta[0] == self._valid_metadata:
+            try: 
+                #version checking
+                if bittensor.__version_as_int__ != int(meta[2].value):
+                    self._deny
+
+                nounce, pubkey, message = meta[1].value.split('bitxx')
+
+                data_time = datetime.strptime(nounce,'%m%d%Y%H%M%S%f')
+                _keypair = self.keypair(ss58_address=pubkey)
+                
+                #checking the time of creation, compared to previous messages
+                if pubkey in self.nounce_dic.keys():
+                    prev_data_time = self.nounce_dic[pubkey]
+                else:
+                    self.nounce_dic[pubkey] = data_time
+                    verification = _keypair.verify(nounce+pubkey,message)
+                    if verification:
+                        return continuation(handler_call_details)
+
+                if data_time - prev_data_time >= timedelta(milliseconds=1):
```

Comment:
I think you need prev_data_time but this does not exist unless the first if statement passes